### PR TITLE
sidecar: Added support for streaming, chunked remote read.

### PR DESCRIPTION
**Pull request by bwplotka**
Fixes: https://github.com/improbable-eng/thanos/issues/488

* [x] Upstream remote stream server side to Prometheus. (in progress: https://github.com/prometheus/prometheus/pull/5703)
 
Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

**Comment by bwplotka** [comment]
https://github.com/prometheus/prometheus/pull/5703 should be able to merge today/following days. Can we have a review on this one as well then, so we can get advantage of it from next Thanos release? (: 

@GiedriusS @povilasv @brancz @krasi-georgiev @domgreen  PTAL if you have time.

**Comment by bwplotka** [comment]
Somehow I keep either OOMing or remote receiver not being in result 
```
query_test.go:82: unexpected error: unexpected result size, expected 4; result: up{instance="localhost:9091", job="prometheus", prometheus="prom-9091", replica="0"} => 1 @[1566231373.376]
up{instance="localhost:9091", job="prometheus", prometheus="prom-ha", replica="0"} => 1 @[1566231373.376]
up{instance="localhost:9091", job="prometheus", prometheus="prom-ha", replica="1"} => 1 @[1566231373.376]
```
```
        level=error name=remote-write-receive-1 ts=2019-08-19T16:16:14.339571874Z caller=config.go:207 component=config-watcher msg="failed to read configuration file" err="open /tmp/fileSD511518933/data/receiveFile1/hashrings.json: no such file or directory" path=/tmp/fileSD511518933/data/receiveFile1/hashrings.json

```
Investigating.

File: pkg/store/prometheus.go [hunk 1]
```diff
@@ -162,12 +165,110 @@ func (p *PrometheusStore) Series(r *storepb.SeriesRequest, s storepb.Store_Serie
...
+
+			thanosChks := make([]storepb.AggrChunk, len(series.Chunks))
+
+			for i, chk := range series.Chunks {
+				thanosChks[i] = storepb.AggrChunk{
+					MaxTime: chk.MaxTimeMs,
+					MinTime: chk.MinTimeMs,
+					Raw: &storepb.Chunk{
+						Data: chk.Data,
+						Type: storepb.Chunk_Encoding(chk.Type - 1),
```
Comment (GiedriusS) [inline pkg/store/prometheus.go]: Just a quick question: why are we subtracting 1 from `chk.Type`?

Comment (bwplotka) [reply pkg/store/prometheus.go]: Will add comment (:

File: pkg/store/prometheus.go [hunk 2]
```diff
@@ -249,47 +381,26 @@ func (p *PrometheusStore) promSeries(ctx context.Context, q prompb.Query) (*prom
 		return nil, errors.Wrap(err, "unable to create request")
 	}
 	preq.Header.Add("Content-Encoding", "snappy")
-	preq.Header.Set("Content-Type", "application/x-protobuf")
-	preq.Header.Set("X-Prometheus-Remote-Read-Version", "0.1.0")
+	preq.Header.Set("Content-Type", "application/x-stream-protobuf")
+	preq.Header.Set("X-Prometheus-Remote-Read-Version", "0.2.0")
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply pkg/store/prometheus.go]: It does not matter much, but somehow client always stated 0.1.0.

Can remove totally.

File: pkg/store/prometheus.go [hunk 3]
```diff
@@ -162,12 +165,110 @@ func (p *PrometheusStore) Series(r *storepb.SeriesRequest, s storepb.Store_Serie
...
+	contentType := httpResp.Header.Get("Content-Type")
+	if strings.HasPrefix(contentType, "application/x-protobuf") {
+		return p.handleSampledPrometheusResponse(s, httpResp, queryPrometheusSpan, externalLabels)
+	}
+
+	if !strings.HasPrefix(contentType, "application/x-streamed-protobuf; proto=prometheus.ChunkedReadResponse") {
+		return errors.Errorf("not supported remote read content type: %s", contentType)
+	}
+
+	level.Debug(p.logger).Log("msg", "started handling ReadRequest_STREAMED_XOR_CHUNKS streamed read response.")
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply pkg/store/prometheus.go]: done.

File: pkg/store/prometheus.go [hunk 4]
```diff
@@ -162,12 +165,110 @@ func (p *PrometheusStore) Series(r *storepb.SeriesRequest, s storepb.Store_Serie
...
+			return errors.Wrap(err, "next proto")
+		}
+
+		if len(res.ChunkedSeries) != 1 {
+			level.Warn(p.logger).Log("msg", "Prometheus ReadRequest_STREAMED_XOR_CHUNKS returned non 1 series in frame", "series", len(res.ChunkedSeries))
+		}
+
+		framesNum++
+		for _, series := range res.ChunkedSeries {
+			hash := func() string {
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply pkg/store/prometheus.go]: Rewritten this indeed

**Comment by bwplotka** [comment]
@krasi-georgiev The merging multi series frame are tested in proxy code here: https://github.com/thanos-io/thanos/blob/master/pkg/store/proxy_test.go#L70

**Comment by bwplotka** [comment]
Started discussion on the StoreAPI Series API invariant aka `multi series frame` here: https://github.com/thanos-io/thanos/issues/1464

**Comment by bwplotka** [comment]
Changed this accordingly to https://github.com/thanos-io/thanos/issues/1464 decision. CC @krasi-georgiev

**Pull request merged**